src/wallet/rpcdump.cpp

```diff
@@ -1482,3 +1482,263 @@ UniValue importmulti(const JSONRPCRequest& mainRequest)
+                have_privkeys = false;
+            }
+        }
+
+        // If private keys are enabled, abort if private keys are not provided and not in the wallet
+        if (!pwallet->IsWalletFlagSet(WALLET_FLAG_DISABLE_PRIVATE_KEYS) && !have_privkeys) {
+            throw JSONRPCError(RPC_WALLET_ERROR, "Cannot import descriptor without private keys to a wallet with private keys enabled");
+        }
+
+        // Range check
+        int64_t range_start = 0, range_end = 1;
+        if (!parsed_desc->IsRange() && data.exists("range")) {
+            throw JSONRPCError(RPC_INVALID_PARAMETER, "Range should not be specified for an un-ranged descriptor");
+        } else if (parsed_desc->IsRange()) {
+            if (!data.exists("range")) {
+                throw JSONRPCError(RPC_INVALID_PARAMETER, "Descriptor is ranged, please specify the range");
+            }
+            auto range = ParseDescriptorRange(data["range"]);
+            range_start = range.first;
+            range_end = range.second;
+            if (range_start < 0 || (range_end >> 31) != 0 || range_end - range_start >= 1000000) {
+                throw JSONRPCError(RPC_INVALID_PARAMETER, "Invalid descriptor range specified");
+            }
+        }
+
+        WalletDescriptor w_desc(std::move(parsed_desc), timestamp, range_start, range_end, range_start);
+
+        // Check if the wallet already contains the descriptor
+        if (pwallet->HasWalletDescriptor(w_desc)) {
+            throw JSONRPCError(RPC_WALLET_ERROR, strprintf("The wallet already has descriptor '%s'", descriptor));
+        }
+
+        // Add descriptor to the wallet
+        auto spk_manager = pwallet->AddWalletDescriptor(w_desc, keys, label);
+        if (spk_manager == nullptr) {
+            throw JSONRPCError(RPC_WALLET_ERROR, strprintf("Could not add descriptor '%s'", descriptor));
+        }
+
+        // Set descriptor as active if necessary
+        if (active) {
+            pwallet->SetActiveScriptPubKeyMan(spk_manager->GetID(), output_type, internal);
+        }
+
+        result.pushKV("success", UniValue(true));
+    } catch (const UniValue& e) {
+        result.pushKV("success", UniValue(false));
+        result.pushKV("error", e);
+    } catch (...) {
+        result.pushKV("success", UniValue(false));
+
+        result.pushKV("error", JSONRPCError(RPC_MISC_ERROR, "Missing required fields"));
+    }
+    if (warnings.size()) result.pushKV("warnings", warnings);
+    return result;
+}
+
+UniValue importdescriptors(const JSONRPCRequest& main_request) {
+    // Acquire the wallet
+    std::shared_ptr<CWallet> const wallet = GetWalletForJSONRPCRequest(main_request);
+    CWallet* const pwallet = wallet.get();
+    if (!EnsureWalletIsAvailable(pwallet, main_request.fHelp)) {
+        return NullUniValue;
+    }
+
+    if (main_request.fHelp || main_request.params.size() < 1 || main_request.params.size() > 2)
+        throw std::runtime_error(
```

Comment:
There is a new thing we do for this check and exception that is part of RPCHelpMan. Most other RPCs do this already, so you should check those for an example.